### ISSUE TITLE
Problem with OrdScope() and SetDeleted( TRUE )

### ISSUE THREAD

**Issue by cpyrgas** [X-Sharp/XSharpPublic#120]
In the following code OrdKeyCount() returns the same value no matter if SetDeleted( TRUE ) or SetDeleted( TRUE ) is used and the dbf has deleted records

```
FUNCTION Start() AS VOID
	LOCAL cDBF, cPfad, cDriver, cIndex AS STRING 
	LOCAL aFields, aValues AS ARRAY 
	LOCAL i AS DWORD                  
	
	cPfad := "C:\test\"
	
	cDriver := RddSetDefault ( "DBFCDX" )  
	
	cDBF := cPfad + "Foo"
	cIndex := cPfad + "Foox" 
	
	FErase ( cIndex + IndexExt() )	
	
	aFields := { { "LAST" , "C" , 20 , 0 } } 
	aValues := { "a1" , "a2", "a3" , "a4" , "a5" , "a6" ,"g1", "g2" , "g3" , "g4" , "g5" , "o1" , "o2" , "o3" , "o4" , "o5" , "u1", "u2","u3" , "u4" }	
	
	? DbCreate( cDBF , AFields)
	? DbUseArea( ,,cDBF , , TRUE )		
	
	FOR i := 1 UPTO ALen ( aValues )
		DbAppend() 
		FieldPut ( 1 , aValues [ i ] ) 
		IF InList ( Upper ( aValues [ i ] ) , "G1" , "G2" , "O5" ) 
			? "DBDelete()" , DbDelete() 			
		ENDIF 
	NEXT 
	
	? DbCreateOrder ( "ORDER1" , cIndex , "upper(LAST)" , { || Upper ( _Field->LAST) } )
	
	? DbSetOrder ( 1 )  
	?
	
	// ---------
	
	SetDeleted ( TRUE ) 
	
	OrdScope ( TOPSCOPE, "G" ) 
	OrdScope ( BOTTOMSCOPE, "G" ) 
// DbGoTop() 	
	
	? OrdKeyCount() // 3 ok
	
	SetDeleted ( FALSE ) 
//	DbGoTop() 		
	
	? OrdKeyCount() // 5 ok
	
	// -------------
	
	OrdScope ( TOPSCOPE, NIL ) 
	OrdScope ( BOTTOMSCOPE, NIL ) 
	
	// If TOPSCOPE and BOTTOMSCOPE is not equal 
	// OrdKeyCount() seems to ignore a SetDeleted( TRUE ) ?
	
	OrdScope ( TOPSCOPE, "A" ) 
	OrdScope ( BOTTOMSCOPE, "G" ) 
	
	?  
	SetDeleted( TRUE ) 	
	? OrdKeyCount()   // 11	 but must show 9 

	SetDeleted( FALSE ) 
	? OrdKeyCount()   // 11	ok
	//------------  
	
	DbCloseAll() 
RETURN
```

**Comment by RobertvanderHulst** [comment]
Chris,
The same code also shows 11 with SetDeleted(TRUE) in VO for DBFCDX and DBFNTX.
OrdKeyCount is not supposed to respect SetDeleted(). If it did it would make the function VERY slow for large files, because the Deleted flag is not stored in the index, so for all keys in the index the DBF would have to be read.

**Comment by Karl1155** [comment]
Robert,

but why does the X# OrdKeycount() respect the setdeleted() true/false setting in this case ?

```
FUNCTION OrdKeyCountTest() AS VOID
	LOCAL cDBF, cPfad, cDriver, cIndex AS STRING 
	LOCAL aFields, aValues AS ARRAY 
	LOCAL i AS DWORD                  
	
	cPfad := "C:\test\"
	
	cDriver := RddSetDefault ( "DBFCDX" )  
	
	cDBF := cPfad + "Foo"
	cIndex := cPfad + "Foox" 
	
	FErase ( cIndex + IndexExt() )	
	
	aFields := { { "LAST" , "C" , 20 , 0 } } 
	aValues := { "a1" , "a2", "a3" , "a4" , "a5" , "a6" ,"g1", "g2" , "g3" , "g4" , "g5" , "o1" , "o2" , "o3" , "o4" , "o5" , "u1", "u2","u3" , "u4" }	
	
	? DbCreate( cDBF , AFields)
	? DbUseArea( ,,cDBF , , TRUE )		
	
	FOR i := 1 UPTO ALen ( aValues )
		DbAppend() 
		FieldPut ( 1 , aValues [ i ] ) 
		IF InList ( Upper ( aValues [ i ] ) , "G1" , "G2" , "O5" ) 
			? "DBDelete()" , DbDelete() 			
		ENDIF 
	NEXT 
	
	? DbCreateOrder ( "ORDER1" , cIndex , "upper(LAST)" , { || Upper ( _Field->LAST) } )
	
	? DbSetOrder ( 1 )  
	?
	
	// ---------
	
	SetDeleted ( TRUE )
	
	? OrdKeyCount()  , "SetDeleted (TRUE)"  //  17
	
	SetDeleted ( FALSE ) 

	? OrdKeyCount() , "SetDeleted (FALSE)"   	// 20
	
	DbCloseAll() 
RETURN	

```

i´m aware of that VO returns in both cases 20, and not 17 and 20.

regards
Karl-Heinz

**Comment by RobertvanderHulst** [comment]
Fixed with 5b86db9a2f883c0251cd6b91afab434b2cfa59b2
